r0.1
.Dd Mar 7, 2021
.Dt R0 1
.Os
.Th R0 1
.Sh NAME
r0 \- interactive raw editor
.Sh SYNOPSIS
.Nm r0
.Op Fl hnv
.Op Fl i Ar script
.Ar file ...
.Sh DESCRIPTION
Simple and minimal hexadecimal editor based on radare concepts
.Pp
The program arguments are:
.Pp
.Bl -tag -width Fl
.It Fl h
Show help message
.It Fl v
Show version information
.It Fl n
Do not be verbose
.It Fl i Ar script
Interpret a script file as r0 commands.
.Pp
Same as "r0 file < script"
.El
.PP
.Sh COMMANDS
.Pp
.Bl -tag -width Fl
.It s[+-addr] 
seek to relative or absolute address
.Pp
> s 0x400  # seek to offset 0x400
.Pp
> s+20     # seek 20 bytes forward
.It b[+-size] 
change block size
.Pp
> b 5      # set block size to 5 bytes
.Pp
> b +128   # increase block size in 128 bytes
.It w[hex|"str"]
write hexpair or string
.Pp
> w 001122       # write 0x00 0x11 0x22
.Pp
> w"Hello World" # write "Hello World" string without quotes
.It /[hex|"str"]
search hexpair or string
.Pp
> /894722       # search bytes 0x89 0x47 0x22
.Pp
> /"lib"        # search string 'lib'
.It x[size]
hexdump
.Pp
> x 256    # dump 256 bytes in hexadecimal
.It X[size]
hexpair dump
.Pp
> X 0x100  # dump 256 bytes in hexpairs
.It p[fmt]
print formatted current block ('p' for help)
.Pp
ob/wW/dD/qQ  byte (oct,hex), word, dword, qword (lil, big endian)
.Pp
i/I/f/F      int32 (lil, big), float (lil, big)
.Pp
s/S          short int16 (lil, big)
.Pp
z/Z          zero-terminatted string (ascii, wide-ascii)
.Pp
./:/*        skip 1 or 4 chars, repeat last format instead of cycle
.It r[-[num]]
truncate or -remove N bytes
.Pp
> r 1024   # resize file to 1024 bytes
.Pp
> r-10     # remove 10 bytes from current seek and shrink file
.It .[file]
interpret file
.Pp
> .script # run file 'script'
.It <[file]
load file in current seek
.Pp
> <curblock.bin  # load blocksize bytes from file
.It >[file]
dump current block to file
.Pp
> >curblock.bin  # dump current block to a file
.It !cmd
run shell command
.Pp
The following environment variables are defined when found in the command string:
.Pp
FILE is the filename of the current working file
.Pp
BLOCK points to the temporal filename with the current block dump
.Pp
OFFSET and XOFFSET are decimal and hexadecimal values of the current seek
.Pp
> !cp $BLOCK $FILE-$XOFFSET.curblock # save current block
.It ?expr
calculate numeric expression
.Pp
> ?1024
.Pp
0x400 1024 02000

.It q
quit

.El
.Sh SEE ALSO
.Pp
.Xr r0v(1)
.Sh AUTHORS
.PP
pancake <@nopcode.org>


r0.c
/* MIT 2009-2021 -- pancake /at/ nopcode /dot/ org */

#define ut64 unsigned long long 
#define ut8 unsigned char

static int earlyquit = 0;
static int verbose = 1;
static int scriptn = 0;
static const char **scripts = 0;
static int cmdn = 0;
static char **cmds = 0;
static ut64 oldseek, curseek = 0LL;
static int obsize, bsize = 256;
int r0_cmd(const char *cmd); // XXX : recursive depenency
#if defined(__WATCOMC__)
#define HAVE_FTRUNCATE 0
#define BUFSZ 1024
#else
#define BUFSZ 128*1024
#endif

ut64 calc(const char *str);
#include "r0.h"
#include "util.c"
#include "calc.c"
#include "cmd.c"

static void r0_slurpin() {
	ut8 buf[BUFSZ];
	for(;;) {
		int len = read(0, buf, sizeof(buf));
		if(len<1) break;
		hexdump(buf, len, 16);
		curseek += len;
	}
}

static int r0_interpret(const char *file) {
	char buf[BUFSZ];
	FILE *fd = fopen(file, "r");
	if(fd != NULL) {
		for(;;) {
			if(fgets(buf, sizeof(buf), fd) == NULL)
				break;
			r0_cmd(buf);
		}
		fclose(fd);
	} else if (file)
		fprintf(stderr, "Cannot open script file '%s'\n", file);
	return 1;
}

int r0_cmd(const char *cmd) {
	const char *arg = cmd+1;
	SKIPSPACES(arg);
	switch(*cmd) {
	case 'q': return 0;
	case ';': case '#': break; // comment
	case '>': return cmd_dump(arg); break;
	case '<': return cmd_load(arg); break;
	case '.': return r0_interpret(arg); break;
	case 's': return cmd_seek(arg); break;
	case 'b': return cmd_bsize(arg); break;
	case '/': return cmd_search(arg); break;
	case 'd': return cmd_disasm(arg); break;
	case 'p': return cmd_print(arg); break;
	case 'r': return cmd_resize(arg); break;
	case 'x': return cmd_hexdump(arg); break;
	case 'X': return cmd_bytedump(arg); break;
	case 'w': return cmd_write(arg); break;
	case '!': return cmd_system(arg); break;
	case 'V': return cmd_system("vr0 $FILE"); break;
	case '?': return cmd_help(arg); break;
	default: fprintf(stderr, "? %s\n", cmd);
	}
	return 1;
}

static int r0_prompt() {
	char *at, *at2, line[BUFSZ];
	if(verbose) {
		printf("[0x%08"LLF"x]> ", curseek);
		fflush(stdout);
	}
	if(fgets(line, sizeof(line), stdin) == NULL)
		return 0;
	line[strlen(line)-1] = '\0';
	if(*line != '!') {
		at = strchr(line, '@');
		oldseek = curseek;
		obsize = bsize;
		if(at) {
			*at = 0;
			at2 = strchr(++at, ':');
			if(at2) {
				*at2 = 0; at2++;
				if(*at2) bsize = (int)calc(at2);
			}
			if(*at) curseek = calc(at);
		}
	}
	at = line;
	SKIPSPACES(at);
	return r0_cmd(at);
}

int r0_open(const char *file) {
	int ret = io_open(file);
	if(ret != -1) {
		oldseek = 0;
		setenv("FILE", file, 1);
		if(scripts)
			for (ret=0;ret<scriptn;ret++)
				r0_interpret(scripts[ret]);
		if(cmds)
			for (ret=0;ret<cmdn;ret++)
				r0_cmd(cmds[ret]);
		if(earlyquit)
			return 0;
		while((ret=r0_prompt())>0) {
			curseek = oldseek;
			bsize = obsize;
		}
		io_close();
	} else if (file)
		fprintf(stderr, "Cannot open '%s'\n", file);
	return ret==-1 ?1:0;
}

static int r0_help() {
	puts("r0 [-qhnv] [-c cmd] [-i script] [-|file ..]");
	return 0;
}

int main(int argc, char **argv) {
	int i, ret = 1;
	argc++;
	cmdn = 0;
	scriptn = 0;
	scripts = malloc(sizeof(const char*)*argc);
	cmds = malloc(sizeof(const char*)*argc);
	if(argc>1 && argv[1]) {
		for(i=1; i<argc; i++) {
			if(argv[i] && argv[i][0]=='-')
				switch(argv[i][1]) {
				case 'q': earlyquit = 1; break;
				case 'i': scripts[scriptn++] = argv[++i]; break;
				case 'c': cmds[cmdn++] = argv[++i]; break;
				case 'n': verbose = 0; break;
				case 'v': puts(VERSION); ret = 0; break;
				case 'h': ret = r0_help(); break;
				case 0x0: r0_slurpin(); ret = 0; break;
			} else {
				if (!argv[i]) break;
				ret = r0_open(argv[i]);
			}
		}
	} else {
		ret = r0_help();
	}
	free(scripts);
        return ret;
}
